pkg/service/storage/ucan_test.go

```diff
@@ -132,6 +132,78 @@ func TestServer(t *testing.T) {
 		})
 	})
 
+	t.Run("repeat blob/allocate for same blob", func(t *testing.T) {
+		space := testutil.RandomDID()
+		size := uint64(rand.IntN(32) + 1)
+		data := testutil.RandomBytes(int(size))
+		digest := testutil.Must(multihash.Sum(data, multihash.SHA2_256, -1))(t)
+		cause := testutil.RandomCID()
+
+		nb := blob.AllocateCaveats{
+			Space: space,
+			Blob: blob.Blob{
+				Digest: digest,
+				Size:   size,
+			},
+			Cause: cause,
+		}
+		cap := blob.Allocate.New(testutil.Alice.DID().String(), nb)
+
+		invokeBlobAllocate := func() result.Result[bdm.AllocateOkModel, ipld.Node] {
+			inv, err := invocation.Invoke(testutil.Service, testutil.Alice, cap, delegation.WithProof(prf))
+			require.NoError(t, err)
+
+			resp, err := client.Execute([]invocation.Invocation{inv}, conn)
+			require.NoError(t, err)
+
+			rcptlnk, ok := resp.Get(inv.Link())
+			require.True(t, ok, "missing receipt for invocation: %s", inv.Link())
+
+			reader := testutil.Must(receipt.NewReceiptReader[bdm.AllocateOkModel, ipld.Node](allocateReceiptSchema))(t)
+			rcpt := testutil.Must(reader.Read(rcptlnk, resp.Blocks()))(t)
+			return rcpt.Out()
+		}
+
+		result.MatchResultR0(invokeBlobAllocate(), func(ok bdm.AllocateOkModel) {
+			fmt.Printf("%+v\n", ok)
+			require.Equal(t, size, uint64(ok.Size))
+			require.NotNil(t, ok.Address)
+		}, func(x ipld.Node) {
+			f := testutil.BindFailure(t, x)
+			fmt.Println(f.Message)
+			fmt.Println(*f.Stack)
+			require.Nil(t, f)
+		})
+
+		// now again without upload
+		result.MatchResultR0(invokeBlobAllocate(), func(ok bdm.AllocateOkModel) {
+			fmt.Printf("%+v\n", ok)
+			require.Equal(t, int64(0), ok.Size)
+			require.NotNil(t, ok.Address)
+		}, func(x ipld.Node) {
+			f := testutil.BindFailure(t, x)
+			fmt.Println(f.Message)
+			fmt.Println(*f.Stack)
+			require.Nil(t, f)
+		})
+
+		// simulate a blob upload
+		err = svc.Blobs().Store().Put(context.Background(), digest, size, bytes.NewReader(data))
+		require.NoError(t, err)
+
+		// now again after upload
+		result.MatchResultR0(invokeBlobAllocate(), func(ok bdm.AllocateOkModel) {
+			fmt.Printf("%+v\n", ok)
+			require.Equal(t, int64(0), ok.Size)
+			require.Nil(t, ok.Address)
+		}, func(x ipld.Node) {
+			f := testutil.BindFailure(t, x)
+			fmt.Println(f.Message)
+			fmt.Println(*f.Stack)
+			require.Nil(t, f)
+		})
+	})
```

Comment:
nit: it may be interesting to add a test for the missing case, the one where the blob has been received but is allocated in a different space, for the sake of completeness

Reply:
Oh yes good shout https://github.com/storacha/storage/pull/20/commits/ab8c03c91712adcc28bebca0d917bbe8aa03749a